src/Dapr.Messaging/PublishSubscribe/PublishSubscribeReceiver.cs

```diff
@@ -0,0 +1,261 @@
+    /// <summary>
+    /// Options allowing the behavior of the receiver to be configured.
+    /// </summary>
+    private readonly DaprSubscriptionOptions options;
+    /// <summary>
+    /// A channel used to decouple the messages received from the sidecar to their consumption.
+    /// </summary>
+    private readonly Channel<TopicMessage> channel = Channel.CreateUnbounded<TopicMessage>();
+    /// <summary>
+    /// A collection of <see cref="TaskCompletionSource"/> used to signal acknowledgement of received messages so a status
+    /// can be sent back to the sidecar indicating what behavior should happen to each.
+    /// </summary>
+    private readonly Dictionary<string, TaskCompletionSource<bool>> acknowledgementTasks = new();
+    /// <summary>
+    /// A semaphore used to ensure thread-safe access to the <see cref="acknowledgementTasks"/> dictionary.
+    /// </summary>
+    private readonly SemaphoreSlim acknowledgementSemaphore = new(1, 1);
+    
+    /// <summary>
+    /// Constructs a new instance of a <see cref="PublishSubscribeReceiver"/> instance.
+    /// </summary>
+    /// <param name="pubsubName">The name of the Dapr pubsub component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Options allowing the behavior of the receiver to be configured.</param>
+    /// <param name="daprClient"></param>
+    internal PublishSubscribeReceiver(string pubsubName, string topicName, DaprSubscriptionOptions options, P.Dapr.DaprClient daprClient)
+    {
+        this.pubsubName = pubsubName;
+        this.topicName = topicName;
+        this.options = options;
+        connectionManager = new ConnectionManager(daprClient);
+    }
+
+    /// <summary>
+    /// Dynamically subscribes to messages on a PubSub topic provided by the Dapr sidecar.
+    /// </summary>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing messages provided by the sidecar.</returns>
+    public IAsyncEnumerable<TopicMessage> SubscribeAsync(CancellationToken cancellationToken)
+    {
+        _ = FetchDataFromSidecar(channel.Writer, cancellationToken);
+        return ReadMessagesFromChannelAsync(channel.Reader, cancellationToken);
+    }
+
+    /// <summary>
+    /// Specifies the action that should be taken on the message after processing it.
+    /// </summary>
+    /// <param name="messageId">The identifier of the message to acknowledge.</param>
+    /// <param name="messageAction">The action to take on the message.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    public async Task AcknowledgeMessageAsync(string messageId, TopicMessageAction messageAction, CancellationToken cancellationToken)
+    {
+        var stream = await connectionManager.GetStreamAsync(cancellationToken);
+        await stream.RequestStream.WriteAsync(new P.SubscribeTopicEventsRequestAlpha1
```

Comment:
Note that gRPC streams must be written to sequentially and the underlying gRPC libraries do not manage such threading.  Writers need to ensure they have exclusive rights to the stream.

Reply:
That's definitely a pro to taking your initial approach. I'm working on rewriting this to adopt that style with some tweaks and will have an update soon.

Reply:
Update in place - described [here](https://github.com/dapr/dotnet-sdk/pull/1346#discussion_r1760465212)